File: anchors/src/main/java/com/effective/android/anchors/LockableAnchor.java
package com.effective.android.anchors;

import android.os.Handler;
import android.util.Log;

public class LockableAnchor {

    private LockListener lockListener;
    private String targetId;
    private Object mLockObject = new Object();
    private Handler handler;
    private boolean unlock = false;

    LockableAnchor(Handler handler){
        this.handler = handler;
    }

    /**
     * 外部监听上锁时机，在监听到上锁之后处理业务逻辑，并完成解锁或者破坏锁行为。
     * @param lockListener
     */
    public void setLockListener(LockListener lockListener){
        this.lockListener = lockListener;
    }

    public String getLockId(){
        return targetId;
    }

    void setTargetTaskId(String id){
        this.targetId = id;
    }

    boolean successToUnlock(){
        return unlock;
    }

    /**
     * 解锁，任务链后续任务继续执行
     */
    public void unlock(){
        handler.post(new Runnable() {
            @Override
            public void run() {
                Logger.d(Constants.LOCK_TAG,Thread.currentThread().getName() + "- unlock( " + targetId + " )");
                Logger.d(Constants.LOCK_TAG,"Continue the task chain...");
                unlockInner();
            }
        });
    }

    synchronized void unlockInner(){
        synchronized (mLockObject) {
            unlock = true;
            mLockObject.notify();
        }
    }

    synchronized void smashInner(){
        synchronized (mLockObject) {
            unlock = false;
            mLockObject.notify();
        }
    }

    /**
     * 破坏锁，任务链后续任务无法继续执行
     */
    public void smash(){
        handler.post(new Runnable() {
            @Override
            public void run() {
                Logger.d(Constants.LOCK_TAG,Thread.currentThread().getName() + "- smash( " + targetId + " )");
                Logger.d(Constants.LOCK_TAG,"Terminate task chain !");
                smashInner();
            }
        });
    }

    /**
     * 上锁
     */
    void lock(){
        Logger.d(Constants.LOCK_TAG,Thread.currentThread().getName() + "- lock( " + targetId + " )");
        try {
            synchronized (mLockObject) {
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        if(lockListener != null){
                            lockListener.lockUp();
                        }
                    }
                });
                mLockObject.wait();
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public interface LockListener{
        void lockUp();
    }
}

File: anchors/src/main/java/com/effective/android/anchors/LockableAnchor.kt
package com.effective.android.anchors

import android.os.Handler
import com.effective.android.anchors.Logger.d

class LockableAnchor internal constructor(private val handler: Handler) {
    private var lockListener: LockListener? = null
    var lockId: String? = null
        private set
    private val mLockObject = Object()
    private var mReleaseListener: ReleaseListener? = null
    private var unlock = false
    /**
     * 外部监听上锁时机，在监听到上锁之后处理业务逻辑，并完成解锁或者破坏锁行为。
     * @param lockListener
     */
    fun setLockListener(lockListener: LockListener) {
        this.lockListener = lockListener
    }

    internal fun setReleaseListener(releaseListener: ReleaseListener) {
        this.mReleaseListener = releaseListener
    }

    internal fun setTargetTaskId(id: String?) {
        lockId = id
    }

    internal fun successToUnlock(): Boolean {
        return unlock
    }

    /**
     * 解锁，任务链后续任务继续执行
     */
    fun unlock() {
        handler.post {
            d(Constants.LOCK_TAG, Thread.currentThread().name + "- unlock( " + lockId + " )")
            d(Constants.LOCK_TAG, "Continue the task chain...")
            unlockInner()
        }
    }

    @Synchronized
    internal fun unlockInner() {
        synchronized(mLockObject) {
            unlock = true
            mLockObject.notify()
            mReleaseListener?.release()
        }
    }

    @Synchronized
    internal fun smashInner() {
        synchronized(mLockObject) {
            unlock = false
            mLockObject.notify()
            mReleaseListener?.release()
        }
    }

    /**
     * 破坏锁，任务链后续任务无法继续执行
     */
    fun smash() {
        handler.post {
            d(Constants.LOCK_TAG, Thread.currentThread().name + "- smash( " + lockId + " )")
            d(Constants.LOCK_TAG, "Terminate task chain !")
            smashInner()
        }
    }

    /**
     * 上锁
     */
    internal fun lock() {
        d(Constants.LOCK_TAG, Thread.currentThread().name + "- lock( " + lockId + " )")
        try {
            synchronized(mLockObject) {
                handler.post(Runnable {
                    lockListener?.lockUp()
                })
                mLockObject.wait()
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    interface LockListener {
        fun lockUp()
    }

    interface ReleaseListener {
        fun release()
    }

}